### PR TITLE
genimage.bbclass: add bmaptool support

### PR DESCRIPTION
**Pull request by Villemoes**
Allow creating a `.bmap` metadata file for use with bmaptool.

### CODE REVIEW

**Review by OnkelUlla — APPROVED** [review]
I utilize bmaptool quite often and therefore really appreciate your enhancement!
Regarding the ``genimage`` alternative ``wic`` Yocto provides automatic treatment with common compressors, see [``image_types.bbclass``](http://git.yoctoproject.org/cgit/cgit.cgi/poky/tree/meta/classes/image_types.bbclass?h=yocto-3.3.1#n261). So you could think of adding something like that on top for even more convenience.

**Comment by Villemoes** [comment]
Absolutely. We're in the process of switching away from wic to genimage, so this was one step towards making it more of a drop-in replacement. I already have some WIP patches for doing the compression, but thought I might as well start out with this small step

**Comment by a3f** [comment]
> We're in the process of switching away from wic to genimage

Having never used wic before, I am very interested in hearing your reasons on why you are migrating to genimage.

**Comment by Villemoes** [comment]
> I utilize bmaptool quite often and therefore really appreciate your enhancement!
> Regarding the `genimage` alternative `wic` Yocto provides automatic treatment with common compressors, see [`image_types.bbclass`](http://git.yoctoproject.org/cgit/cgit.cgi/poky/tree/meta/classes/image_types.bbclass?h=yocto-3.3.1#n261). So you could think of adding something like that on top for even more convenience.

Do you want me to add the compression support as part of this PR, or can this get merged as-is?

**Comment by OnkelUlla** [comment]
> > I utilize bmaptool quite often and therefore really appreciate your enhancement!
> > Regarding the `genimage` alternative `wic` Yocto provides automatic treatment with common compressors, see [`image_types.bbclass`](http://git.yoctoproject.org/cgit/cgit.cgi/poky/tree/meta/classes/image_types.bbclass?h=yocto-3.3.1#n261). So you could think of adding something like that on top for even more convenience.
> 
> Do you want me to add the compression support as part of this PR, or can this get merged as-is?

No, this could be added on top of your current work later.

**Comment by Villemoes** [comment]
> > > I utilize bmaptool quite often and therefore really appreciate your enhancement!
> > > Regarding the `genimage` alternative `wic` Yocto provides automatic treatment with common compressors, see [`image_types.bbclass`](http://git.yoctoproject.org/cgit/cgit.cgi/poky/tree/meta/classes/image_types.bbclass?h=yocto-3.3.1#n261). So you could think of adding something like that on top for even more convenience.
> > 
> > 
> > Do you want me to add the compression support as part of this PR, or can this get merged as-is?
> 
> No, this could be added on top of your current work later.

OK. There's gonna be a conflict with #81 which is about to be merged, once that is in I'll rebase this on top.

File: classes/genimage.bbclass [hunk 1]
```diff
@@ -80,6 +80,7 @@ DEPENDS += "genimage-native"
 GENIMAGE_IMAGE_SUFFIX ?= "img"
 
 GENIMAGE_IMAGE_NAME = "${IMAGE_BASENAME}-${MACHINE}-${DATETIME}"
+GENIMAGE_IMAGE_FULLNAME = "${GENIMAGE_IMAGE_NAME}.${GENIMAGE_IMAGE_SUFFIX}"
```
Comment (ejoerns) [inline classes/genimage.bbclass:83]: This line should not split `GENIMAGE_IMAGE_NAME` from its `[vardepsexclude]` setting. And `GENIMAGE_IMAGE_LINK_NAME` comes after it anyway, too.

**Comment by ejoerns** [comment]
I also appreciate this enhancement. :+1: 

As I know both wic and genimage, knowing the resons for switching to genimage would be quite interesting for me to hear, too ;)

File: classes/genimage.bbclass (hunk 1 above)
Comment (Villemoes) [reply classes/genimage.bbclass]: Fixed, and I've thrown in a GENIMAGE_IMAGE_LINK_FULLNAME as well as I ended up adding another use for that anyway.

**Review by ejoerns — APPROVED** [review]
Despite even oe-core's `image_types.bbclass` does not introduce these abbreviated names for these variable combinations and I am unsure whether these need to be modifiable, I also see that it shortens the code and reduces redundancy a bit, so it is ok for me.

Looks good now. :+1:

**Comment by Villemoes** [comment]
> I also appreciate this enhancement. +1
> 
> As I know both wic and genimage, knowing the resons for switching to genimage would be quite interesting for me to hear, too ;)

I'll just give y'all the short version: wic has baroque and counter-intuitive syntax, seems quite inflexible, and the fact that it is "maintained" in oe-core means that even the simplest of enhancements literally takes months, if not years, to get merged, and when that finally happens you have to wait another half year before you can start relying on that. Contrast with genimage which has more readable syntax (e.g. clearly separating the overall metadata "what partition table type etc." from that of each partition), and when one finds that you miss some features, sending a couple of PRs either gets merged immediately, or you get very constructive feedback and then it gets merged after a couple of rounds. The fact that the tool itself does not live in the yocto meta-layer means I can plug in a .bbappend and have it point at the current genimage master and make use of those features as soon as they've been merged.

**Comment by a3f** [comment]
> Having never used wic before, I am very interested in hearing your reasons on why you are migrating to genimage.

I have since had occasion to try out wic. I needed to recreate an existing partition layout in a BSP without meta-ptx. The partition layout had an out-of-order partition table [which is now supported in genimage](https://github.com/pengutronix/genimage/pull/143) thanks to @Villemoes, but not in wic. So I went back to genimage again. ^^

**Pull request merged**